File: game-server/src/main/java/org/forfun/mmorpg/game/scene/actor/Creature.java
package org.forfun.mmorpg.game.scene.actor;

import jakarta.persistence.Transient;
import org.forfun.mmorpg.game.attribute.AttributeContainer;
import org.forfun.mmorpg.game.attribute.AttributeType;
import org.forfun.mmorpg.game.buff.model.BuffContainer;
import org.forfun.mmorpg.game.scene.model.Life;

/**
 * 生物体，就是会动的场景演员
 *
 */
public abstract class Creature extends SceneActor {

    @Transient
	private BuffContainer buffContainer;

    @Transient
	private AttributeContainer attrContainer = new AttributeContainer();

    @Transient
	private Life life;

	public long getHp() {
		return life.getCurrHp();
	}
	
	public long getAttack() {
		return attrContainer.getAttrValue(AttributeType.Attack);
	}

	public BuffContainer getBuffContainer() {
		return buffContainer;
	}

	public AttributeContainer getAttrContainer() {
		return attrContainer;
	}

	public void setAttrContainer(AttributeContainer attrContainer) {
		this.attrContainer = attrContainer;
	}
	
	public Life getLife() {
		return life;
	}

	public void setLife(Life life) {
		this.life = life;
	}

	public void setBuffContainer(BuffContainer buffContainer) {
		this.buffContainer = buffContainer;
	}

	public int dispatchKey() {
		return mapId + lineId;
	}

	public int dispatchLine() {
		return lineId;
	}
	
	public boolean isDie() {
		return life.getCurrHp() <= 0;
	}

}

File: game-server/src/main/kotlin/org/forfun/mmorpg/game/scene/actor/Creature.kt
package org.forfun.mmmorpg.game.scene.actor

import jakarta.persistence.Transient
import org.forfun.mmorpg.game.attribute.AttributeContainer
import org.forfun.mmorpg.game.attribute.AttributeType
import org.forfun.mmorpg.game.buff.model.BuffContainer
import org.forfun.mmorpg.game.scene.actor.SceneActor
import org.forfun.mmorpg.game.scene.model.Life

/**
 * 生物体，就是会动的场景演员
 */
abstract class Creature : SceneActor() {

    @Transient
    var buffContainer: BuffContainer? = null

    @Transient
    var attrContainer = AttributeContainer()

    @Transient
    var life: Life? = null

    val hp: Long
        get() = life!!.currHp

    val attack: Int
        get() = attrContainer.getAttrValue(AttributeType.Attack).toInt()

    fun dispatchKey(): Int = mapId + lineId

    fun dispatchLine(): Int = lineId

    val isDie: Boolean
        get() = life!!.currHp <= 0
}
